src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -66,14 +70,19 @@ public Task Save(IContainSagaData sagaData, SagaCorrelationProperty correlationP
             var inMemSession = (InMemorySynchronizedStorageSession) session;
             inMemSession.Enlist(() =>
             {
-                if (correlationProperty != SagaCorrelationProperty.None)
+                var lockenTokenKey = $"{sagaData.GetType().FullName}.{(correlationProperty == null ? "None" : correlationProperty.Name)}.{(correlationProperty == null ? "None" : correlationProperty.Value)}";
+                var lockToken = lockers.GetOrAdd(lockenTokenKey, key => new object());
+                lock (lockToken)
                 {
-                    ValidateUniqueProperties(correlationProperty, sagaData);
+                    if (correlationProperty != SagaCorrelationProperty.None)
+                    {
+                        ValidateUniqueProperties(correlationProperty, sagaData);
+                    }
+
+                    data.AddOrUpdate(sagaData.Id,
+                        id => new VersionedSagaEntity(sagaData, lockenTokenKey),
+                        (id, original) => new VersionedSagaEntity(sagaData, lockenTokenKey, original)); // we can never end up here.
```

Comment:
Look at it again and still on the fence. TryAdd would work but TryUpdate is just super clumsy. Assuming I made a mistake while thinking it through and we would hit the Update part of the concurrent dictionary with the current code base it would still work (although lead to suboptimal locking). I'm leaning now to leave it as is